### PR TITLE
SPOI-8561 Record reader example app

### PR DESCRIPTION
**Pull request by yogidevendra**

### CODE REVIEW

**Comment by yogidevendra** [comment]
@amberarrow Could you please review this?

**Comment by yogidevendra** [comment]
Note that this is dependent on https://github.com/apache/apex-malhar/pull/326 which is not yet merged. Thus, this PR is only for the review for now.

**Comment by yogidevendra** [comment]
@amberarrow parent PR apache/apex-malhar#326 is merged now. This PR is for example app for record-reader. Could you please review this?

**Comment by amberarrow** [comment]
Please add a row to the table in the parent directory README.md with a brief description of what the example does (order is alphabetic).

File: tutorials/recordReader/src/test/java/com/example/recordReader/ApplicationTest.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      LocalMode lma = LocalMode.newInstance();
+      Configuration conf = new Configuration(false);
+      conf.addResource(this.getClass().getResourceAsStream("/META-INF/properties.xml"));
+      lma.prepareDAG(new Application(), conf);
+      LocalMode.Controller lc = lma.getController();
+      lc.run(10 * 1000); // runs for 10 seconds and quits
+    } catch (ConstraintViolationException e) {
+      Assert.fail("constraint violations: " + e.getConstraintViolations());
+    }
+  }
```
Comment (amberarrow) [inline tutorials/recordReader/src/test/java/com/example/recordReader/ApplicationTest.java]: Please rewrite test to actually check results as done in the other examples.

Also, the sample data `CustomerTxnData` file is rather large; can the size be reduced ?

**Comment by yogidevendra** [comment]
1. Trimmed `CustomerTxnData` file
2. Added entry in the parent README.md

In the current implementation, output is written to the Console. So, we cannot monitor the output files directly. Are you suggesting to write to fileoutput rather than console? But, earlier in the JIRA discussion we decided for Console output to keep it simple. 

Could you please suggest why checking the output is really important? Just checking if output files exists may not be sufficient if we are really focusing on validating the functionality. Adding validation logic to the example app would make it complex unnecessarily. Thus, IMO it should be OK to skip the check for output files.  

Awaiting for your feedback on this.

**Comment by yogidevendra** [comment]
@amberarrow Updated the PR to incorporate review comments.

**Comment by tweise** [comment]
It will be good to have the unit test verify the result. That will be necessary to automate testing of apps as part of certification.

File: tutorials/recordReader/src/test/java/com/example/recordReader/ApplicationTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  {
+    try {
+      LocalMode lma = LocalMode.newInstance();
+      Configuration conf = new Configuration(false);
+      conf.addResource(this.getClass().getResourceAsStream("/META-INF/properties.xml"));
+      conf.setBoolean("dt.application.RecordReaderExample.operator.console.prop.saveTuples", true);
+      lma.prepareDAG(new Application(), conf);
+      ToConsole console = (ToConsole)lma.getDAG().getOperatorMeta("console").getOperator();
+      LocalMode.Controller lc = lma.getController();
+      lc.run(10 * 1000); // runs for 10 seconds and quits
```
Comment (amberarrow) [inline tutorials/recordReader/src/test/java/com/example/recordReader/ApplicationTest.java]: run(10 * 1000) => runAsync()

**Comment by amberarrow** [comment]
@tweise Commit 1565003 does some verification in the unit test; were you looking for additional verification or just responding to the earlier question from Yogi ?

**Comment by yogidevendra** [comment]
@amberarrow changed as per your comment.

@tweise Awaiting for your answer to the ram's question.

File: tutorials/recordReader/README.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+This example shows how to use `FSRecordReaderModule` to read newline delimited records from a file convert them to 
+plain old java objects (POJO) using `CSVParser`. These POJOs are emitted on stdout using `ConsoleOutputOperator`. 
+
+The properties file `META-INF/properties.xml` shows how to configure the resepective operators.
+
+The application can be run on an actual cluster or in local mode within your IDE by
+simply running the method `ApplicationTest.testApplication()`.
+
+You may tweak this example to connect to some other operator instead of `ConsoleOutputOperator` to achieve functionality
```
Comment (tweise) [inline tutorials/recordReader/README.md]: It's not using ConsoleOutputOperator?

**Comment by tweise** [comment]
Rather than just counting, should the verification not also check that a particular record was correctly parsed?

**Comment by yogidevendra** [comment]
@tweise In my understanding, intent of the example app is do demonstrate how to use a particular operator/feature in question. 

We already have unit tests to verify the individual operators. 

Intricacies like end to end verification of data will be part of certification. But, that can be done outside the app as well. Even if we decide to build the app for verification, that app will be part of QA and not the examples. 

Increasing the complexity in the example app would make it difficult to understand. Hence, I was trying to keep the verification to minimal.  

Awaiting for your feedback on this.

**Comment by tweise** [comment]
Examples not only illustrate the application code. Documentation and testing is equally if not more important. Users need to see and be easily able to adapt the recommended approach for testing their own app instead of "figure yourself". End to end data verification is IMO part of the package and cannot be thrown over the fence. Running the tests can be automated elsewhere.

**Comment by yogidevendra** [comment]
@tweise Updated the PR to do the verification of results. Could you please review this?

**Comment by amberarrow** [comment]
The README file now seems out of sync with the code: it mentions "ToConsole" which no longer exists.
Also, could you remove the 2 comment lines, one with ToConsole and the other with TODO  in ApplicationTest.java ?
Rest looks good.

**Comment by yogidevendra** [comment]
@amberarrow Updated the PR. Could you please review?

**Pull request merged**